### PR TITLE
Extend Document weight attribute interface to use dictionary snapshot

### PR DESCRIPTION
**Pull request by toregge**
(for reproducible lookup) and store enum index for lookup key.

@geirst : please review

### CODE REVIEW

File: searchlib/src/vespa/searchlib/attribute/multinumericpostattribute.h [hunk 1]
```diff
@@ -35,7 +35,9 @@ class MultiValueNumericPostingAttribute
     struct DocumentWeightAttributeAdapter : IDocumentWeightAttribute {
         const MultiValueNumericPostingAttribute &self;
         DocumentWeightAttributeAdapter(const MultiValueNumericPostingAttribute &self_in) : self(self_in) {}
-        virtual LookupResult lookup(const vespalib::string &term) const override final;
+        virtual vespalib::datastore::EntryRef get_dictionary_snapshot() const override final;
```
Comment (baldersheim) [inline searchlib/src/vespa/searchlib/attribute/multinumericpostattribute.h:38]: drop virtual and final and just make the class final.

File: searchlib/src/vespa/searchlib/attribute/multistringpostattribute.h [hunk 2]
```diff
@@ -33,7 +33,9 @@ class MultiValueStringPostingAttributeT
     struct DocumentWeightAttributeAdapter : IDocumentWeightAttribute {
         const MultiValueStringPostingAttributeT &self;
         DocumentWeightAttributeAdapter(const MultiValueStringPostingAttributeT &self_in) : self(self_in) {}
-        virtual LookupResult lookup(const vespalib::string &term) const override final;
+        virtual vespalib::datastore::EntryRef get_dictionary_snapshot() const override final;
```
Comment (baldersheim) [inline searchlib/src/vespa/searchlib/attribute/multistringpostattribute.h:36]: drop virtual and final and just make the class final.

**Review by baldersheim — APPROVED** [review]

**Review by baldersheim — APPROVED** [review]

File: searchlib/src/vespa/searchlib/attribute/i_document_weight_attribute.h [hunk 3]
```diff
@@ -17,11 +19,14 @@ struct IDocumentWeightAttribute
...
-            : posting_idx(posting_idx_in), posting_size(posting_size_in), min_weight(min_weight_in), max_weight(max_weight_in) {}
+        const vespalib::datastore::EntryRef enum_idx;
+        LookupResult() : posting_idx(), posting_size(0), min_weight(0), max_weight(0), enum_idx() {}
+        LookupResult(vespalib::datastore::EntryRef posting_idx_in, uint32_t posting_size_in, int32_t min_weight_in, int32_t max_weight_in, vespalib::datastore::EntryRef enum_idx_in)
+            : posting_idx(posting_idx_in), posting_size(posting_size_in), min_weight(min_weight_in), max_weight(max_weight_in), enum_idx(enum_idx_in) {}
     };
-    virtual LookupResult lookup(const vespalib::string &term) const = 0;
+    virtual vespalib::datastore::EntryRef get_dictionary_snapshot() const = 0;
+    virtual LookupResult lookup(const vespalib::string &term, vespalib::datastore::EntryRef dictionary_snapshot) const = 0;
+    virtual void collect_folded(vespalib::datastore::EntryRef enum_idx, vespalib::datastore::EntryRef dictionary_snapshot, const std::function<void(vespalib::datastore::EntryRef)>& callback) const = 0;
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/attribute/i_document_weight_attribute.h:29]: Please add a description on what this function is supposed to do.

**Review by geirst — APPROVED** [review]
👍

**Review by geirst — APPROVED** [review]
👍

**Pull request merged**